### ISSUE TITLE
[Performance]: A performance optimization in the Scheduler regarding pad_spec_decode

### ISSUE THREAD

**Issue by zzzzwwjj** [vllm-project/vllm#55798]
### Misc discussion on performance

The code for `pad_spec_decode` in the current Scheduler is written as follows:

```py
# Pad new decode requests to uniform spec decoding size to
# preserve full cudagraph for this step.
# Not for diffusion where draft tokens can't be padded.
if (
    (self.num_spec_tokens > 0 and self.dynamic_sd_lookup is None)
    and self.num_sampled_tokens_per_step > 0
    and num_new_tokens == 1
    and (scheduled_running_reqs and not prefill_scheduled)
):
    padded_num_tokens = 1 + self.num_spec_tokens
    # Pad only when there is room for the sampled token(s).
    if (
        num_computed_tokens
        + padded_num_tokens
        + self.num_sampled_tokens_per_step
        <= self.max_model_len
    ):
        if padded_num_tokens > request_token_budget:
            # Prefer to not schedule than schedule un-padded.
            break
        num_new_tokens = padded_num_tokens
        pad_spec_decode = True
```

There is a non empty judgment of `scheduled_running_reqs` in the branch judgment condition, but it will result in the first request being unable to pad when `scheduled_running_reqs` is empty, thus unable to enter graph mode.

Why we need to judge `scheduled_running_reqs` is not None?

### Your current environment (if you think it is necessary)

```text
The output of `python collect_env.py`
```

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by twu3202** [comment]
The snippet in the report matches `releases/v0.29.0`. On `main` this condition was already relaxed by #55126 (merged 2026-09-04), which is what the report is asking for:

```python
# releases/v0.29.0
and (scheduled_running_reqs and not prefill_scheduled)

# main @ fbf51c7
and not prefill_scheduled
and (scheduled_running_reqs or num_computed_tokens > 0)
```

On the intent of the original check: the block sits in the waiting-queue loop, and `scheduled_running_reqs` is filled by the running-queue loop earlier in the same step, so it gates requests entering from the waiting queue that need exactly one token.

I measured this on an RTX 6000 Ada with `Qwen/Qwen3-0.6B`, ngram `num_speculative_tokens=3`, prefix caching on. Instrumentation only, recording both predicates in one run without changing behaviour. One priming request, then the identical prompt six times with the engine otherwise idle:

| | count |
|---|---|
| requests reaching the branch | 6 |
| `releases/v0.29.0` predicate true | 0 |
| `main` predicate true | 6 |
| actually padded (to `1 + 3` tokens) | 6 |

So on the release branch none of the six repeats is padded, and on `main` all six are. Scope is one scheduling step per affected request, not one per token; I did not measure the resulting step time, only the scheduling decision.

One detail worth adding: the trigger is wider than the PD case in #55126's title. `get_computed_blocks` caps a lookup at `num_tokens - 1` so the last token is recomputed for logits, which means a plain full prefix-cache hit on a repeated prompt also arrives as a one-token waiting request. Any deployment with prefix caching and gaps between requests hits this, no disaggregation needed. Preemption does not: a preempted request re-prefills from zero, and the branch is never reached.

#55126 is not on `releases/v0.29.0`. If a back-port is wanted I am happy to open one.
